lib/kdtree.cpp

```diff
@@ -3,304 +3,497 @@
-    if (tris.size() == 0) {
-        return nullptr;
+    TreeNode* left() const {
+        assert(is_inner());
+        return reinterpret_cast<TreeNode*>(left_or_triangle_ids_);
     }
 
-    // to few triangles -> terminate
-    if (tris.size() <= 3) {
-        return new Node(tris);
+    TreeNode* right() const {
+        assert(is_inner());
+        return right_;
     }
 
-    // box too small -> no need to split further -> terminate
-    if (box.surface_area() == 0) {
-        return new Node(tris);
+    TriangleId num_tris() const {
+        assert(is_leaf());
+        return flags_;
     }
 
-    float min_cost;
-    Axis plane_ax;
-    float plane_pos; // plane
-    TriangleIds ltris, rtris;
-    std::tie(min_cost, plane_ax, plane_pos, ltris, rtris) =
-        find_plane_and_classify(tris, box);
-
-    // automatic termination
-    // remove lambda factor from cost again, otherwise we may stuck in an
-    // empty space split forever
-    if (COST_INTERSECTION * tris.size() * lambda(ltris.size(), rtris.size()) <
-        min_cost) {
-        return new Node(tris);
+    TriangleId* triangle_ids() const {
+        assert(is_leaf());
+        return reinterpret_cast<TriangleId*>(left_or_triangle_ids_);
     }
 
-    Box lbox, rbox;
-    std::tie(lbox, rbox) = box.split(plane_ax, plane_pos);
+    // the last 3 values are reserved for axis description
+    static constexpr TriangleId MAX_ID =
+        std::numeric_limits<TriangleId>::max() - 3;
+
+private:
+    static constexpr TriangleId FLAG_AXIS_X = MAX_ID + 1;
+    static constexpr TriangleId FLAG_AXIS_Y = MAX_ID + 2;
+    static constexpr TriangleId FLAG_AXIS_Z = MAX_ID + 3;
+
+    // If flags_ in [0, MAX_ID], then this node is a leaf node and this
+    // member is the number of triangles in the leaf. Otherwise, this node
+    // is an inner node. In this case, flags is in [FLAG_AXIS_X,
+    // FLAG_AXIS_Y, FLAG_AXIS_Z] and describes the splitting axis.
+    TriangleId flags_; // 4 bytes
+    // Used only in an inner node
+    float split_pos_; // 4 bytes
+    // Depending whether this node is an inner node or a leaf, this member
+    // is a pointer to the left child or resp. the array containing the
+    // triangles.
+    void* left_or_triangle_ids_; // 8 bytes
+    // Used only in an inner node
+    TreeNode* right_; // 8 bytes
+                      // == 24 bytes
+};
 
-    Node* left = build(std::move(ltris), lbox);
-    Node* right = build(std::move(rtris), rbox);
-    assert(left || right);
+/**
+ * Self contained implementation of Algorithm 4 from:
+ *
+ * "On building fast kd-Trees for Ray Tracing, and on doing that in O(N log N)"
+ * by Ingo Wald and Vlastimil Havran
+ * [WH06]
+ *
+ * Builds up a kd-tree in O(N log^2 N), however the resulting kd-tree is the
+ * same as if it were built with Algorithm 5.
+ *
+ * TODO: Replace by Algorithm 5.
+ */
+class KDTreeBuildAlgorithm {
+public:
+    KDTreeBuildAlgorithm(const Triangles& triangles) : triangles_(&triangles) {}
```

Comment:
I like `HavranWaldAlgorithm` or `KDTreeHavranWald`. Since mostly everything is an algorithm. And I thought you like free functions ;).